presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -246,10 +246,7 @@ public PlanWithProperties visitEnforceSingleRow(EnforceSingleRowNode node, Conte
         @Override
         public PlanWithProperties visitAggregation(AggregationNode node, Context context)
         {
-            HashSet<Symbol> partitioningRequirement = new HashSet<>(node.getGroupingSets().get(0));
-            for (int i = 1; i < node.getGroupingSets().size(); i++) {
-                partitioningRequirement.retainAll(node.getGroupingSets().get(i));
-            }
+            Set<Symbol> partitioningRequirement = ImmutableSet.copyOf(node.getGroupingKeys());
```

Comment:
Sorry, I don't think I explained my point well. I'm saying that a partial aggregation should never introduce exchanges below itself due to partitioning requirements. Partial aggregations should deal with whatever data they get and *then* partition so that the *finals* always get the right groups.